src/DotVVM.Framework/Compilation/Binding/ExpressionBuildingVisitor.cs

```diff
@@ -317,16 +360,49 @@ protected override Expression VisitLambda(LambdaBindingParserNode node)
             var body = Visit(node.BodyExpression);
 
             ThrowOnErrors();
-            return Expression.Lambda(body, lambdaParameters);
+            return CreateLambdaExpression(body, lambdaParameters, typeInferenceData.Type);
         }
 
         protected override Expression VisitLambdaParameter(LambdaParameterBindingParserNode node)
         {
-            if (node.Type == null)
+            if (node.Type == null && node.ResolvedType == null)
                 throw new BindingCompilationException($"Could not infer type of parameter.", node);
 
-            var parameterType = Visit(node.Type).Type;
-            return Expression.Parameter(parameterType, node.Name.ToDisplayString());
+            if (node.ResolvedType != null)
+            {
+                // Type was not specified but was infered
+                return Expression.Parameter(node.ResolvedType, node.Name.ToDisplayString());
+            }
+            else
+            {
+                // Type was specified and needs to be obtained from binding node
+                var parameterType = Visit(node.Type).Type;
+                return Expression.Parameter(parameterType, node.Name.ToDisplayString());
+            }
+        }
+
+        private Expression CreateLambdaExpression(Expression body, ParameterExpression[] parameters, Type? delegateType)
+        {
+            if (delegateType != null)
+            {
+                if (delegateType.Name.StartsWith("Action"))
```

Comment:
This can potentially match things we don't want - I'd check namespace and also presence of a backtick.